Add converters to the capsule API

lpw25 opened the pull request:
This PR extends the capsule API with a new `Converter.t` type that doesn't cross contention and, when `uncontended`, can freely wrap and unwrap `Data.t` values. They are made available by a `Capsule.access` function, which many of the `Capsule.Data` functions could be implemented in terms of. They should make some patterns slightly easier to write.

Converters replace the `expose` function and the use of global `Password.t`s for converting values, which allows us to safely mark `Password.t`s as crossing contention which makes it easier to nest their usage. Separating these two roles also helps to clarify the relationship between capsules and threads.

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
+type initial
+
+(** A converter for the initial capsule *)
+val initial : initial Converter.t
```
Perhaps we should have an `initial Password.t` instead?

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
+
+    (* CR layouts v5: this should have layout [void], but
+       [void] can't be used for function argument and return types yet. *)
+    type 'k t : value mod global portable many unique
```
Why is this `portable`?

lpw25 replied:
As discussed offline, only because it safely can be not because it is actually useful. I prefer to reserve portability for controlling things that are like functions, and using contention for controlling things that are like mutable state, and I picture the converters as acting more like mutable state.

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
 
+val access :
+  'k Password.t @ local
+  -> ('k Converter.t -> 'a @ portable) @ local portable
```
I guess this shouldn't be able to leak a converter without marking it contended.

lpw25 replied:
Right, we aren't using locality to constrain converters. Instead they are treated like mutable data that is part of the capsule. So if they cross the capsule boundary they become contended and can no longer be used.

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
-        and modes prevent retaining the [t] after releasing the mutex. This
-        guarantees that uncontended access to the capsule is only granted
-        to a single domain at once. *)
+    type 'k t : value mod portable many unique uncontended
```
Same for contention.

dkalinichenko-js replied:
Locality is not sufficient to guarantee non-shared access in presence of `fork_join`.

lpw25 replied:
Yeah, with `fork_join` we will be relying on the `unyielding/yielding` modes to prevent passwords from being used by two threads at the same time. You can think of a password as an effect handler, and `yielding` is how we prevent those from being used inside a `fork_join` too.

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -108,9 +152,21 @@ module Data : sig
-    exception Contended of exn @@ contended
-    (** If a function accessing the contents of the capsule raises an exception,
-        it is wrapped in [Contended] to avoid leaking access to the data. *)
+    val wrap :
```
Oh, I see, any way to send a converted across domains should mark it `contended`, which will prevent its usage.

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
+       [void] can't be used for function argument and return types yet. *)
+    type 'k t : value mod global portable many unique
+    (** ['k t] is the type of "converters" that can wrap and unwrap
+        [Data.t] values. An [uncontended] ['k t] indicates that ['k] is the
```
Does that mean `assert_current : 'a Converter.t -> 'b Converter.t -> ('a, 'b) Type.eq` is sound?

lpw25 replied:
Yeah, I guess it does. Maybe we should include it?

lpw25 replied:
I went ahead and added it

dkalinichenko-js commented on ocaml/otherlibs/stdlib_alpha/capsule.mli:
```diff
@@ -40,25 +40,68 @@
 
 end
 
+val access :
```
`access` should be `@@ portable`.

dkalinichenko-js commented:
Is it correct to think of `('k Converter.t @ local -> ...) @ portable` as a modality indexed by the capsule `'k` that has access to all variables from the background context guarded by `(..., 'k) Data.t` (in addition to all `contended` variables)?

lpw25 commented:
I think the `local` isn't needed, but yes I think you can think of that as some form of modality.

lpw25 commented:
I've renamed `Converter.t` to `Access.t` which is a bit better. I've also updated the API to account for reader/writer locks. Notes that, rather than add a new `Access.Shared.t` to represent shared access, we just use having an `Access.t` at the `shared` mode to represent shared access.

dkalinichenko-js commented:
Happy with the updated PR. To be sure, we can't have `Password.t @ shared` because passwords should mode-cross contention, right?
